File: 2016/13_AMazeofTwistyLittleCubicles/cubicles.ts
// ======================================================================
// A Maze of Twisty Little Cubicles
//   Advent of Code 2016 Day 13 -- Eric Wastl -- https://adventofcode.com
//
// TypeScript implementation by Dr. Dean Earl Wright III
// ======================================================================

// ======================================================================
//                           c u b i c l e s . t s
//
// A solver for the Advent of Code 2016 Day 13 problem
// ======================================================================

// ----------------------------------------------------------------------
//                                                                 import
// ----------------------------------------------------------------------

// ----------------------------------------------------------------------
//                                                                  types
// ----------------------------------------------------------------------
type Location = number;
type Directions = 0 | 1 | 2 | 3;

type NextCube = [number, number, number, number];
interface Cube {
  location: Location;
  next: NextCube;
  previous: Location;
  fromStart: number;
}
type Cubes = Record<Location, Cube>;

type XY = [number, number];

// ----------------------------------------------------------------------
//                                                              constants
// ----------------------------------------------------------------------
const locationConstant = 1000;
const startX = 1;
const startY = 1;
const finishX = 31;
const finishY = 39;
const badNext: Location = -1;
const north: Directions = 0;
const east: Directions = 1;
const south: Directions = 2;
const west: Directions = 3;

const dirNames = ['North', 'East', 'South', 'West'];
const deltas = [-1, locationConstant, 1, -locationConstant];
const reverse = [south, west, north, east];

// ======================================================================
//                                                               Cubicles
// ======================================================================

export class Cubicles {
  // Object for A Maze of Twisty Little Cubicles
  text: string[];

  part2: boolean;

  start: Location;

  finish: Location;

  cubes: Cubes;

  favorite: number;

  wall: number;

  furthest: number;

  constructor(text: string[], part2 = false) {
    // Create a Cubicles object

    // 1. Set the initial values
    this.text = text === undefined ? [] : text;
    this.part2 = part2 === undefined ? false : part2;
    this.start = Cubicles.xyToLocation(startX, startY);
    this.finish = Cubicles.xyToLocation(finishX, finishY);
    this.cubes = {};
    this.cubes[this.start] = this.initialLocation();
    this.favorite = 0;
    this.wall = 100;
    this.furthest = 200;
    if (this.part2) this.furthest = 50;

    // 2. Process text (if any)
    if (this.text.length !== 0) {
      this.favorite = parseInt(this.text[0], 10);
    }
  }

  static initializeLocation(loc: Location): Cube {
    const [x, y] = Cubicles.locationToXY(loc);
    const nextCube: NextCube = [NaN, NaN, NaN, NaN];
    if (x === 0) nextCube[west] = badNext;
    if (y === 0) nextCube[north] = badNext;
    return {
      location: loc,
      next: nextCube,
      previous: NaN,
      fromStart: 0,
    };
  }

  initialLocation(): Cube {
    return Cubicles.initializeLocation(this.start);
  }

  static xyToLocation(x: number, y: number): Location {
    return x * locationConstant + y;
  }

  static locationToXY(loc: Location): XY {
    return [Math.trunc(loc / locationConstant), loc % locationConstant];
  }

  isItCube(loc: Location): boolean {
    if (loc in this.cubes) return true;
    const [x, y] = Cubicles.locationToXY(loc);
    if (x < 0 || y < 0) return false;
    if (x > this.wall || y > this.wall) return false;
    const sum = x * x + 3 * x + 2 * x * y + y + y * y + this.favorite;
    const bits = sum.toString(2).replace(/0/g, '').length;
    return bits % 2 === 0;
  }

  stepsTo(verbose = false, limit = 0): number {
    if (verbose) console.log(`Finding steps from ${this.start} to ${this.finish}, limit=${limit}`);
    // 1. Start with nothing at the start
    let result = 99999999;
    if (this.furthest > 0) result = this.furthest;
    let currentLoc = this.start;
    let steps = 0;
    // 2. Loop until we run out of options or take too long.
    while (!Number.isNaN(currentLoc)
    && (limit === 0 || steps < limit)) {
      steps += 1;
      let nextLoc = this.cubes[currentLoc].previous;
      if (currentLoc === this.finish) {
        if (Number.isNaN(result) || this.cubes[currentLoc].fromStart < result) {
          result = this.cubes[currentLoc].fromStart;
        }
        if (verbose) console.log(`Reached finished ${this.finish} steps=${result}`);
        if (verbose) console.log(`Backtracking from ${currentLoc} to ${nextLoc} due to finding finish`);
      } else if (this.cubes[currentLoc].fromStart >= result) {
        if (verbose) console.log(`Backtracking from ${currentLoc} to ${nextLoc} due to reashing result limit`);
      } else {
        nextLoc = this.nextCube(currentLoc);
        if (nextLoc === badNext) {
          nextLoc = this.cubes[currentLoc].previous;
          if (verbose) console.log(`Backtracking from ${currentLoc} to ${nextLoc} due to getting badNext`);
        } else if (verbose) {
          console.log(`Moving from ${currentLoc} to ${nextLoc}`);
        }
      }
      currentLoc = nextLoc;
    }
    return result;
  }

  nextCube(loc: Location): Location {
    // 0. Prexisting conditions
    if (loc === badNext) return badNext;
    if (!(loc in this.cubes)) return badNext;
    const cube = this.cubes[loc];
    // 1. look for a unexplored direction
    for (let direction = 0; direction < dirNames.length; direction += 1) {
      if (Number.isNaN(cube.next[direction])) {
        // 2. Haven't been this way from here
        const nextLoc = Cubicles.deltaLoc(loc, direction);
        // 3. Is it a wall or cube;
        if (this.isItCube(nextLoc)) {
          // 3a. A cube
          this.cubes[loc].next[direction] = nextLoc;
          // 4. If we never been here before, create it
          if (!(nextLoc in this.cubes)) {
            const newCube = Cubicles.initializeLocation(nextLoc);
            newCube.fromStart = cube.fromStart + 1;
            newCube.next[reverse[direction]] = loc;
            newCube.previous = loc;
            this.cubes[nextLoc] = newCube;
            return nextLoc;
          }
          // 5. Been here before
          if (this.cubes[nextLoc].fromStart > cube.fromStart + 1) {
            // console.log(`From ${loc} going ${dirNames[direction]} been to ${nextLoc} before; Changing steps from ${this.cubes[nextLoc].fromStart} to ${cube.fromStart + 1}`);
            this.cubes[nextLoc].next = [NaN, NaN, NaN, NaN];
            this.cubes[nextLoc].fromStart = cube.fromStart + 1;
            this.cubes[nextLoc].next[reverse[direction]] = loc;
            this.cubes[nextLoc].previous = loc;
            return nextLoc;
          }
        } else {
          // 3b. A wall
          this.cubes[loc].next[direction] = badNext;
        }
      }
    }
    return badNext;
  }

  static deltaLoc(loc: Location, direction: number): Location {
    return loc + deltas[direction];
  }

  solution(verbose = false, limit = 0): number {
    if (verbose) console.log(`solution: ${limit}`);
    if (this.part2) {
      this.finish = 9999999;
      this.stepsTo(verbose, limit);
      return Object.keys(this.cubes).length;
    }
    return this.stepsTo(verbose, limit);
  }

  partOne(verbose = false, limit = 0): number {
    // Returns the solution for part one

    return this.solution(verbose, limit);
  }

  partTwo(verbose = false, limit = 0): number {
    // Returns the solution for part two

    // 1. Return the solution for part two
    return this.solution(verbose, limit);
  }
}

// ======================================================================
// end                      c u b i c l e s . t s                     end
// ======================================================================

File: 2016/13_AMazeofTwistyLittleCubicles/cubicles.py
# ======================================================================
# A Maze of Twisty Little Cubicles
#   Advent of Code 2016 Day 13 -- Eric Wastl -- https://adventofcode.com
#
# Python implementation by Dr. Dean Earl Wright III
# ======================================================================

# ======================================================================
#                         c u b i c l e s . p y
# ======================================================================
"A solver for the Advent of Code 2016 Day 13 puzzle"

# ----------------------------------------------------------------------
#                                                                 import
# ----------------------------------------------------------------------

# ----------------------------------------------------------------------
#                                                              constants
# ----------------------------------------------------------------------
OPEN = 0
WALL = 1

DIRS = [(1, 0), (0, 1), (-1, 0), (0, -1)]

# ======================================================================
#                                                               Cubicles
# ======================================================================


class Cubicles(object):   # pylint: disable=R0902, R0205
    "Object for A Maze of Twisty Little Cubicles"

    def __init__(self, text=None, part2=False):

        # 1. Set the initial values
        self.part2 = part2
        self.text = text
        self.maze = {(1, 1): OPEN}
        self.number = 0
        if text:
            self.number = int(text[0])

    def which(self, x, y):
        "Determine if the area is a wall or open"
        if x < 0 or y < 0:
            return WALL
        if (x, y) in self.maze:
            return self.maze[(x, y)]
        value = self.number + x * x + 3 * x + 2 * x * y + y + y * y
        knt = bin(value).count("1")
        result = knt % 2
        self.maze[(x, y)] = result
        #  print("x=%d, y=%d, v=%d, k=%d r=%d" % (x, y, value, knt, result))
        return result

    def shortest(self, x, y):
        "Determine the shortest path from (1, 1) to (x, y)"
        # 1. Set up maximum steps
        if self.part2:
            shortest = 51
        else:
            shortest = 999
        visited = {(1, 1): 0}
        # 2. Initialize goal and search
        goal = (x, y)
        backtrack = [((1, 1), self.directions(1, 1))]
        # 3. Loop until there is no more to search
        while len(backtrack) > 0:
            loc, dirs = backtrack.pop()
            steps = visited[loc]
            # 4. If goal is reached,
            if loc == goal:
                shortest = min(steps, shortest)
            # 5. If it is worth continuing from here ...
            if steps + 1 < shortest:
                # 6. Determine where to go next
                goto = dirs.pop()
                # 7. If there are more exits from here, we still need to search from those
                if len(dirs) > 0:
                    backtrack.append((loc, dirs))
                # 8. If this place is new (or we got here faster), add it for further searching
                if goto not in visited or (goto in visited and steps + 1 < visited[goto]):
                    backtrack.append((goto, self.directions(goto[0], goto[1])))
                    visited[goto] = steps + 1
        # 9. Return the answer for the question for part 1 or 2
        if self.part2:
            return len(visited)
        else:
            return shortest

    def directions(self, x, y):
        "Determine the directions from the indicated space"
        result = []
        for delta in DIRS:
            new_x = x + delta[0]
            new_y = y + delta[1]
            if self.which(new_x, new_y) == OPEN:
                result.append((new_x, new_y))
        return result

    def part_one(self, verbose=False, limit=0):
        "Returns the solution for part one"

        # 0. Precondition axioms
        assert verbose in [True, False]
        assert limit >= 0

        # 1. Return the solution for part one
        return self.shortest(31, 39)

    def part_two(self, verbose=False, limit=0):
        "Returns the solution for part two"

        # 0. Precondition axioms
        assert verbose in [True, False]
        assert limit >= 0

        # 1. Return the solution for part two
        return self.shortest(999, 999)


# ----------------------------------------------------------------------
#                                                  module initialization
# ----------------------------------------------------------------------
if __name__ == '__main__':
    pass

# ======================================================================
# end                      c u b i c l e s . p y                     end
# ======================================================================
